### PR TITLE
Update ALE for ROM fixes and snapshot/restore of state

### PR DESCRIPTION
**Pull request by shelhamer**
ALE development has gone on from the point that atari-py forked. While atari-py forked from ALE without keeping it's history, the diff can be inspected by comparing `atari-py/atari_py/ale_interface/src` and `ale/src`. Among the various fixes and improvements two stand out for the purposes of gym:

- ROM fixes that alter the env semantics(!)
- proper snapshot/restore of the emulator state including pseudorandomness (`cloneSystemState()` and `restoreSystemState()`)

This is required to address OpenAI/gym#402. I'm happy to carry out the gym change to expose the snapshot/restore functionality next.

**Do not merge until we double-check that nothing breaks and decide that we're happy with the new ROM semantics. Once merged, gym should bump the env versions when the new atari-py is incorporated.**

### CODE REVIEW

**Comment by shelhamer** [comment]
Note that I almost certainly blew away improves by Trevor Blackwell—I'll review the history and spare them from this diff.

**Comment by shelhamer** [comment]
Right now this includes new ALE output on when making an env. I'll most likely disable it, or better yet switch it to being a verbose option.

```
A.L.E: Arcade Learning Environment (version 0.5.1)
[Powered by Stella]
Use -help for help screen.
Warning: couldn't load settings file: ./ale.cfg
Game console created:
  ROM file:  /Users/shelhamer/h/desk/atari/atari-py/atari_py/atari_roms/pong.bin
  Cart Name: Video Olympics (1978) (Atari)
  Cart MD5:  60e0ea3cbe0913d39803477945e9e5ec
  Display Format:  AUTO-DETECT ==> NTSC
  ROM Size:        2048
  Bankswitch Type: AUTO-DETECT ==> 2K

Running ROM file...
Random seed is 436853534
[2017-04-12 23:31:47,088] Making new env: PongNoFrameskip-v3
```

**Comment by shelhamer** [comment]
Note to self: double-check stellarc/ale.cfg

**Comment by shelhamer** [comment]
Alright, this should be ready. I did a few simple checks of the environments, but more testing and review might be a good idea before merge.

**Comment by joschu** [comment]
@nottombrown could you try running our RL algorithm baselines with this branch?

**Comment by nottombrown** [comment]
Bumped the version and am testing now using PPO (because it's fast)
```
 ॐ  python -c 'import pkg_resources;  print(pkg_resources.require("atari-py")[0].version)'
0.0.22.dev0
```

**Comment by nottombrown** [comment]
Results
image

**Comment by joschu** [comment]
Awesome -- the rare monotonic improvement. Let's merge!

**Comment by shelhamer** [comment]
@nottombrown I'd expect slight improvement in Qbert and Breakout because of minor alterations of the ROMs in 199fdf3 (which fix a terminal condition and reduce the minimal action set respectively), and otherwise equivalent results.

However, *seeds are not preserved* across this change because of a switch in the ALE RNG in a5241ce6b74211171a48e82c29197839d748255f; see https://github.com/mgbellemare/Arcade-Learning-Environment/commits/master/src/emucore/Random.cxx for more history. There is still determinism for the new seed/RNG scheme, but it will give different random numbers so it's all the more important to bump the Atari envs to v4.

**Comment by shelhamer** [comment]
Thanks @nottombrown for catching the version bump!

File: atari_py/ale_interface/src/external/TinyMT/tinymt32.c [hunk 1]
```diff
@@ -60,6 +60,12 @@ static void period_certification(tinymt32_t * random) {
  * @param seed a 32-bit unsigned integer used as a seed.
  */
 void tinymt32_init(tinymt32_t * random, uint32_t seed) {
+    // MGB: These values were apparently drawn from the aether. This isn't good for reproducibility.
+    // I will set them to some bunk values.
+    random->mat1 = 4753849;
+    random->mat2 = 3231259923;
+    random->tmat = 614784120;
+
     int i;
```
Comment (Kojoley) [inline atari_py/ale_interface/src/external/TinyMT/tinymt32.c]: Only GNU extension allows declaring variables in the middle of the block.

**Pull request merged**